### ISSUE TITLE
Question about the dataset and label.txt

### ISSUE THREAD

**Issue by ambipomyan** [facebookresearch/dinov2#142]
Hi, I have a question about how to handle the dataset and label.txt when I want to train the model. I am a bit confused when I see the settings for labels since the training is of unsupervised learning, then how I handle the data structure of the dataset and contexts of label.txt for training?

I guess the labels should not do anything for the unsupervised training and now I just put the inputs, which are images with no label, into separated folders and recognize the name of the folders as the name of class in order to fit the format of dinov2 inputs. In this way, the name of class and the images in the folders are definitely not matching. Will this work for training?

Thank you in advance!

dataset
```
- ROOT
   |-- train
   |   |-- folder0
   |   |    |-- folder0_01.jpeg
   |   |    `-- ...
   |   |-- folder1
   |   |    |-- folder1_01.jpeg
   |   |    `-- ...
   |   `-- ...
   |-- val
   |-- test
   `-- label.txt
```
label.txt
```
folder0, folder0
folder1, folder1
...
```

**Comment by yyyyyyfs** [comment]
As far as I know，training or finetuning this model is not self-supervised. It is also need label，so if you want to train your dataset， you should generate your labels.txt, and modify the data_loader to adapt your dataset.

**Comment by ddstone** [comment]
> As far as I know，training or finetuning this model is not self-supervised. It is also need label，so if you want to train your dataset， you should generate your labels.txt, and modify the data_loader to adapt your dataset.

So this project does not support to pre-train dinov2 using custom dataset of no label, with the so-called `Discriminative Self-supervised Pre-training` in paper.

**Comment by TheoMoutakanni** [comment]
Hello,
The training of DINOv2 is fully unsupervised.
We don't use labels during the training but we use them during the validation process where we fit a linear layer on top of our frozen features to check that the model is training well.
You shouldn't need labels to launch a training of DINOv2, our datasets have labels for evaluation but if you don't have labels you can return something like '0' or 'np.nan'.

Look at this file to see what you need for a minimal dataset:
https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/extended.py

To make a new dataset without labels, you would have to make a class that inherits `ExtendedVisionDataset` and create functions for `get_image_data` (should return an image as an array), `get_target` (in your case this should return 0 or nan) and `__len__` (this should returns the lenght of the dataset).

You can take inspiration from the ImageNet dataset class in https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/image_net.py

But keep in mind that you don't need to copy all functions in the ImageNet dataset.

**Comment by ddstone** [comment]
> Hello, The training of DINOv2 is fully unsupervised. We don't use labels during the training but we use them during the validation process where we fit a linear layer on top of our frozen features to check that the model is training well. You shouldn't need labels to launch a training of DINOv2, our datasets have labels for evaluation but if you don't have labels you can return something like '0' or 'np.nan'.
> 
> Look at this file to see what you need for a minimal dataset: https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/extended.py
> 
> To make a new dataset without labels, you would have to make a class that inherits `ExtendedVisionDataset` and create functions for `get_image_data` (should return an image as an array), `get_target` (in your case this should return 0 or nan) and `__len__` (this should returns the lenght of the dataset).
> 
> You can take inspiration from the ImageNet dataset class in https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/image_net.py
> 
> But keep in mind that you don't need to copy all functions in the ImageNet dataset.

Appreciate your rely, i've launched the training with dataset without labels, by mocking the `image_net.py`. Thank you!

**Comment by HDL-YD** [comment]
> > 您好，DINOv2 的训练完全无人监督。我们在训练过程中不使用标签，但我们在验证过程中使用它们，在冻结特征之上拟合一个线性层，以检查模型是否训练良好。启动 DINOv2 的训练不需要标签，我们的数据集有用于评估的标签，但如果你没有标签，你可以返回类似“0”或“np.nan”的内容。
> > 查看此文件以了解最小数据集所需的内容：https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/extended.py
> > 要创建一个没有标签的新数据集，你必须创建一个类来继承和创建函数（应该将图像作为数组返回），（在你的例子中，这应该返回0或nan）和（这应该返回数据集的长度）。`ExtendedVisionDataset``get_image_data``get_target``__len__`
> > 您可以从 ImageNet 数据集类中获取灵感 https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/image_net.py
> > 但请记住，您不需要复制 ImageNet 数据集中的所有函数。
> 
> 感谢您的依赖，我已经通过嘲笑 .谢谢！`image_net.py`
I've also recently trained a backbone network with my own dataset. You can share your dataset structure and the changed loading dataset code


> > Hello, The training of DINOv2 is fully unsupervised. We don't use labels during the training but we use them during the validation process where we fit a linear layer on top of our frozen features to check that the model is training well. You shouldn't need labels to launch a training of DINOv2, our datasets have labels for evaluation but if you don't have labels you can return something like '0' or 'np.nan'.
> > Look at this file to see what you need for a minimal dataset: https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/extended.py
> > To make a new dataset without labels, you would have to make a class that inherits `ExtendedVisionDataset` and create functions for `get_image_data` (should return an image as an array), `get_target` (in your case this should return 0 or nan) and `__len__` (this should returns the lenght of the dataset).
> > You can take inspiration from the ImageNet dataset class in https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/image_net.py
> > But keep in mind that you don't need to copy all functions in the ImageNet dataset.
> 
> Appreciate your rely, i've launched the training with dataset without labels, by mocking the `image_net.py`. Thank you!

I've also recently trained a backbone network with my own dataset. You can share your dataset structure and the changed loading dataset code

**Comment by HDL-YD** [comment]
> > Hello, The training of DINOv2 is fully unsupervised. We don't use labels during the training but we use them during the validation process where we fit a linear layer on top of our frozen features to check that the model is training well. You shouldn't need labels to launch a training of DINOv2, our datasets have labels for evaluation but if you don't have labels you can return something like '0' or 'np.nan'.
> > Look at this file to see what you need for a minimal dataset: https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/extended.py
> > To make a new dataset without labels, you would have to make a class that inherits `ExtendedVisionDataset` and create functions for `get_image_data` (should return an image as an array), `get_target` (in your case this should return 0 or nan) and `__len__` (this should returns the lenght of the dataset).
> > You can take inspiration from the ImageNet dataset class in https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/image_net.py
> > But keep in mind that you don't need to copy all functions in the ImageNet dataset.
> 
> Appreciate your rely, i've launched the training with dataset without labels, by mocking the `image_net.py`. Thank you!

I've also recently trained a backbone network with my own dataset. Can you share your dataset structure and the changed loading dataset code?

**Comment by csaroff** [comment]
I have a fork that trains a dinov2 model with an [arbitrary unlabeled dataset](https://github.com/csaroff/dinov2/blob/main/dinov2/data/datasets/recursive_image_dataset.py) using [skypilot](https://skypilot.readthedocs.io/en/latest/).

https://github.com/csaroff/dinov2/tree/main/sky

Should be easy enough to edit the config to [pull down your dataset](https://github.com/csaroff/dinov2/blob/main/sky/vitl-p14-bs1024-ep10.yaml#L25) instead of mine

Only caveat is that resuming training from a checkpoint isn't working(re-running training will overwrite old checkpoints instead of resuming) so you either can't use spot instances, or you'll have to debug that particular issue.

**Comment by shiyongde** [comment]
same problem

**Comment by ayushnangia** [comment]
did you figure out how to just pretrain by initializing the weights from the model weights given? @csaroff @shiyongde

**Comment by PMRS-lab** [comment]
> > Hello, The training of DINOv2 is fully unsupervised. We don't use labels during the training but we use them during the validation process where we fit a linear layer on top of our frozen features to check that the model is training well. You shouldn't need labels to launch a training of DINOv2, our datasets have labels for evaluation but if you don't have labels you can return something like '0' or 'np.nan'.
> > Look at this file to see what you need for a minimal dataset: https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/extended.py
> > To make a new dataset without labels, you would have to make a class that inherits `ExtendedVisionDataset` and create functions for `get_image_data` (should return an image as an array), `get_target` (in your case this should return 0 or nan) and `__len__` (this should returns the lenght of the dataset).
> > You can take inspiration from the ImageNet dataset class in https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/image_net.py
> > But keep in mind that you don't need to copy all functions in the ImageNet dataset.
> 
> Appreciate your rely, i've launched the training with dataset without labels, by mocking the `image_net.py`. Thank you!

May I ask how to launch the training with dataset without labels, by mocking the `image_net.py`. Thank you very much!

**Comment by PMRS-lab** [comment]
> I have a fork that trains a dinov2 model with an [arbitrary unlabeled dataset](https://github.com/csaroff/dinov2/blob/main/dinov2/data/datasets/recursive_image_dataset.py) using [skypilot](https://skypilot.readthedocs.io/en/latest/).
> 
> https://github.com/csaroff/dinov2/tree/main/sky
> 
> Should be easy enough to edit the config to [pull down your dataset](https://github.com/csaroff/dinov2/blob/main/sky/vitl-p14-bs1024-ep10.yaml#L25) instead of mine
> 
> Only caveat is that resuming training from a checkpoint isn't working(re-running training will overwrite old checkpoints instead of resuming) so you either can't use spot instances, or you'll have to debug that particular issue.

May I ask how to launch the training with dataset without labels, by mocking the image_net.py. Thank you very much!

**Comment by PMRS-lab** [comment]
> > Hello, The training of DINOv2 is fully unsupervised. We don't use labels during the training but we use them during the validation process where we fit a linear layer on top of our frozen features to check that the model is training well. You shouldn't need labels to launch a training of DINOv2, our datasets have labels for evaluation but if you don't have labels you can return something like '0' or 'np.nan'.
> > Look at this file to see what you need for a minimal dataset: https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/extended.py
> > To make a new dataset without labels, you would have to make a class that inherits `ExtendedVisionDataset` and create functions for `get_image_data` (should return an image as an array), `get_target` (in your case this should return 0 or nan) and `__len__` (this should returns the lenght of the dataset).
> > You can take inspiration from the ImageNet dataset class in https://github.com/facebookresearch/dinov2/blob/main/dinov2/data/datasets/image_net.py
> > But keep in mind that you don't need to copy all functions in the ImageNet dataset.
> 
> Appreciate your rely, i've launched the training with dataset without labels, by mocking the `image_net.py`. Thank you!

May I ask how to launch the training with dataset without labels, by mocking the image_net.py. Thank you very much!

**Comment by csaroff** [comment]
@PMRS-lab I have instructions for launching training in the [README](https://github.com/csaroff/dinov2/blob/main/sky/README.md).

In my case, I subclassed [ExtendedVisionDataset](https://github.com/csaroff/dinov2/blob/main/dinov2/data/datasets/recursive_image_dataset.py) rather than messing with image_net.py directly.

Really hope this helps. Feel free to [create an issue](https://github.com/csaroff/dinov2/issues/new) if you have any questions about the fork!

**Comment by PMRS-lab** [comment]
> @PMRS-lab I have instructions for launching training in the [README](https://github.com/csaroff/dinov2/blob/main/sky/README.md).
> 
> In my case, I subclassed [ExtendedVisionDataset](https://github.com/csaroff/dinov2/blob/main/dinov2/data/datasets/recursive_image_dataset.py) rather than messing with image_net.py directly.
> 
> Really hope this helps. Feel free to [create an issue](https://github.com/csaroff/dinov2/issues/new) if you have any questions about the fork!

Thank you very much! I replaced your file recursive_image_dataset.py with the original file extended.py and ran the train /run/train/train.py file, but still couldn't train. Can you please explain your changes in detail？ I really need to fine tune my dataset to achieve feature extraction of images instead of recognition and classification tasks. Thank you again.

**Comment by csaroff** [comment]
> I replaced your file recursive_image_dataset.py with the original file extended.py and ran the train /run/train/train.py file, but still couldn't train.

You don't need to replace recursive_image_dataset.py. It's referenced in [configs/train/vitl14_ep10.yaml](https://github.com/csaroff/dinov2/blob/main/dinov2/configs/train/vitl14_ep10.yaml#L9) which is referenced from the [skypilot config](https://github.com/csaroff/dinov2/blob/main/sky/vitl-p14-bs1024-ep10.yaml#L43).

Did you try launching with the skypilot config based on the [README instructions](https://github.com/csaroff/dinov2/blob/main/sky/README.md)?

> Can you please explain your changes in detail?

I subclassed ExtendedVisionDataset based on [the instructions above](https://github.com/facebookresearch/dinov2/issues/142#issuecomment-1722973008) in order train a model without any labels. I also added a skypilot configuration to run training on A100 nodes.

If you're looking for additional information about [my changes](https://github.com/csaroff/dinov2/compare/273a960671f9a7e9588ecb768c1f661ebde495cc...csaroff:dinov2:main) that isn't already in the README, you're welcome to [create a github issue](https://github.com/csaroff/dinov2/issues/new) or [start a discussion](https://github.com/csaroff/dinov2/discussions/new/choose) on my fork. Otherwise we risk creating a lot of noise in this thread 📣😃

Hope this helps!
